### ISSUE TITLE
ietf-yang-library module version support Inconsistencies

### ISSUE THREAD

**Issue by iffy50** [CESNET/libyang#1052]
Hi,

Since the update to support both the 2016-04-09 (1.0) and 2019-04-09(1.1) versions of the ietf-yang-library modules.

Here's the relevant entries from the capabilities announcement:
```
<capability>urn:ietf:params:netconf:capability:yang-library:1.0?revision=2019-01-04&amp;module-set-id=34</capability>
<capability>urn:ietf:params:netconf:capability:yang-library:1.1?revision=2019-01-04&amp;content-id=34</capability>
```
Both instances are showing the 2019-01-04 revision date, but with two different YANG version numbers. Surely the 1.0 capability should be for the 2016-04-09 version of ietf-yang-library?

The installed version:
```
# sysrepoctl -l |grep library
ietf-yang-library          | 2019-01-04 | I     | root:root | 666         |            |
```

Further to that, the schema for the returned xml actually seems to be using the 2016-04-09 schema. The schema and datastore lists are under netconf-state, not under yang-library as in the 2019 model:
```
 <netconf-state xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-monitoring">
    <capabilities>
      <capability>urn:ietf:params:netconf:base:1.0</capability>
 ....
    </capabilities>
    <datastores>
      <datastore>
        <name>running</name>
      </datastore>
      <datastore>
        <name>startup</name>
      </datastore>
      <datastore>
        <name>candidate</name>
      </datastore>
    </datastores>
    <schemas>
      <schema>
        <identifier>ietf-yang-metadata</identifier>
        <version>2016-08-05</version>
        <format>yang</format>
        <namespace>urn:ietf:params:xml:ns:yang:ietf-yang-metadata</namespace>
        <location>NETCONF</location>
      </schema>
      <schema>
        <identifier>ietf-yang-metadata</identifier>
        <version>2016-08-05</version>
        <format>yin</format>
        <namespace>urn:ietf:params:xml:ns:yang:ietf-yang-metadata</namespace>
        <location>NETCONF</location>
      </schema>
...
```
So, there seems to be some problems in the versions that are being installed, advertised and the schema version that is actually in use.

**Comment by michalvasko** [comment]
Hi,
okay, this in not completely simple to explain so I will just tell you why it is implemented the way it is without much regard to what you wrote to be less confusing.

So, there is the old YANG library [RFC](https://tools.ietf.org/html/rfc7895) with module revision `2016-06-21` and top-level container `modules-state`. Also, the YANG [RFC](https://tools.ietf.org/html/rfc7950#section-5.6.4) says the following
```
   The server also MUST advertise the following capability in the
   <hello> message (line breaks and whitespaces are used for formatting
   reasons only):

     urn:ietf:params:netconf:capability:yang-library:1.0?
       revision=<date>&module-set-id=<id>

   The parameter "revision" has the same value as the revision date of
   the "ietf-yang-library" module implemented by the server.  This
   parameter MUST be present.
```
So far so good.

Then there is the new YANG library [RFC](https://tools.ietf.org/html/rfc8525). It specifies the same module in revision `2019-01-04` and new top-level container `yang-library`. However, it __also keeps the definition__ of `modules-state` except it is marked as `deprecated`. The new NMDA [RFC](https://tools.ietf.org/html/rfc8526#page-4) says
```
   The server MUST advertise the following capability in the <hello>
   message (line breaks and whitespace are used for formatting reasons
   only):

     urn:ietf:params:netconf:capability:yang-library:1.1?
       revision=<date>&content-id=<content-id-value>

   The parameter "revision" has the same value as the revision date of
   the "ietf-yang-library" module implemented by the server.  This
   parameter MUST be present.

   ...

   This document updates Section 5.6.4 of [RFC7950] to allow servers to
   advertise the capability :yang-library:1.1 instead of
   :yang-library:1.0 and to implement the subtree "/yang-library"
   [RFC8525] instead of "/modules-state".
```

Now for the conclusion of all these specifications. When upgrading to the new `ietf-yang-library` module, we have did exactly what the RFC says, replaced 1.0 capability with 1.1, which resulted in broken compatibility and this issue. Fine, I have tried to fix this by making it as interoperable as possible while not breaking any specifications. So, *libnetconf2* advertises both capabilities, which can already be considered not covered the specs. However, both capabilities explicitly mention that the `revision-date` must be of the __implemented module__ by the server, so the new revision `2019-01-04` MUST be used in both cases. Also, the server cannot implement 2 revisions of the same module. Hence, I believe there is nothing more to be done in *libnetconf2* and any clients expecting the old `modules-state` data tree should probably (explained further) check only for the `yang-library:1.0` capability without expecting any specific revision of the module.

Some more notes. *libyang* (so *libnetconf2* as well) generates both the old deprecated and the new data tree. I am not positive this was really expected (notice the use of *replace* in the new YANG library RFC). It is possible the deprecated tree was meant only for clients, so that they are able to load data from NETCONF servers implementing the old revision of the module. Also, I have tried advertising `yang-library:1.0` with the old revision in *libnetconf2*. The result was that *netopeer2-cli* then requested this old `ietf-yang-library` revision from the server, which failed, of course, because the module was not really implemented or available. This is just another reason why the old revision cannot be advertised if it is not actually implemented.

To summarize, I believe most of the things discussed are in a gray area of the specifications and some clarification on the *netmod* mailing list could help.

Regards,
Michal

**Comment by iffy50** [comment]
Thanks for the explanation. Looking back through the output I'm getting, that makes a lot more sense now.

Then It seems that the reason that we have a problem is not the version of the module as I originally thought. 

For the newer yang-library container module the client (NSO) is expecting to see entries in the ietf-library:module-set:schema and ieft-library:module-set:datastore lists, but these are not present in the returned config. We get:
```
  <yang-library xmlns="urn:ietf:params:xml:ns:yang:ietf-yang-library">
    <module-set>
      <name>complete</name>
      <import-only-module>
        <name>ietf-yang-metadata</name>
        <revision>2016-08-05</revision>
        <namespace>urn:ietf:params:xml:ns:yang:ietf-yang-metadata</namespace>
      </import-only-module>
      <module>
...
```
But the expected response should also contain something like:
```
   <schema>
     <name>complete</name>
     <module-set>complete</module-set>
   </schema>
   <datastore>
     <name xmlns:ds="urn:ietf:params:xml:ns:yang:ietf-datastores">ds:candidate</name>
     <schema>complete</schema>
   </datastore>
```

**Comment by michalvasko** [comment]
Hi,
okay, that makes sense. It has slipped our attention because no one had a problem with it until now but the server should properly advertise all its supported datastores. So, if you update all the projects (*libyang*, *sysrepo*, *libnetconf2*, and *netopeer2*) it should work. `ietf-datastores` now needs to be implemented and that required some change in every project.

Regards,
Michal

**Comment by michalvasko** [comment]
So now only one of the capabilities is advertised and it is possible in *sysrepo* to choose what `ietf-yang-library` revision to use. It was more work than I anticipated but no matter, updating *libyang*, *libnetconf2*, and *sysrepo* is required. There is just one minor problem that if the old revision is used then technically NMDA is not supported (supporting the new revision is a requirement of NMDA) even though NMDA support in *sysrepo* is not optional. But I do not think this is a serious problem.

Regards,
Michal
